「lsコマンドを作る4」の課題としてlsコマンドに-l オプションを追加

cellotak opened the pull request:
「[lsコマンドを作る4](https://bootcamp.fjord.jp/practices/224#learning-Status)」の課題としてlsコマンドに-l オプションを追加しました。

## 要件
lsコマンド全体の要件については下記参照↓
[lsコマンドを作る](https://bootcamp.fjord.jp/pages/ls-command#requirements)

-l オプションについてはOS標準のものに合わせるとのこと

## 実行結果
### 作成したlsコマンドの実行結果
上から順に
1. オプション無
1. -lオプション有
で実行しています。


![image.png](https://bootcamp.fjord.jp/rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBek12QXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--7fd25af6e8ec10308726715964599af85558da5e/image.png)




## OS標準のlsコマンドの実行結果
同様に上から順に
1. オプション無
1. -lオプション有
で実行しています。

![image.png](https://bootcamp.fjord.jp/rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBelF2QXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--d7d8e6c7621ecc1af37c56f9483f32b4c213672f/image.png)

## rubocopのチェック
rubocopのチェックはパスしています。
![image.png](https://bootcamp.fjord.jp/rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBeXd2QXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--174abb7cd9267d8f527ba27d3b81cb60d0a8a361/image.png)


## 必須要件かわからない点
OS標準のlsコマンドは各項目が幅揃えされているのですが、この動作は必須要件でしょうか。
この部分については[こちらのコメント](https://github.com/cellotak/ruby-practices/pull/9/files#r1318199169)に詳しく記載させて頂いてます。
![image](https://github.com/cellotak/ruby-practices/assets/65953037/22bf7f03-8a38-4498-ab17-186ba7753aa0)



## その他補足
- 「-aオプションや-rオプションは実装しなくてよいです。」とありましたが、特に他のオプションがデグレードする要素がなかったため、オプションは残っており、また併用することができます。
- -lオプション無しの時、平仮名や漢字にすると、左揃えでなくなる問題には手をつけていません。
- ディレクトリ名は指定できますが、ファイル名の指定や複数ディレクトリの指定には対応していません。(どこかのタイミングで対応したいとは思っています。)
- lsコマンドにaliasを設定しており、lsと打つとディレクトリ名は青色になるなど、標準とは異なる表示がされてしまうので、\lsという形で実行しています。
- OS標準のlsコマンドとはソートのルールが違うようで、出力順は異なっています。

cellotak committed:
File.statで情報をとってくる部分を実装中

cellotak committed:
ls.rbに追加する-lオプションのため、fetch_file_namesメソッドを変更

fetch_file_namesメソッドをfetch_file_detailsにして、ファイル名を含むファイルの詳細情報が格納されたハッシュを配列にして返すようにしました。
詳細情報としてctimeとsizeのみしかとってきていないですが、今後追加予定。
outputメソッドは修正していないので、とりあえずoutputsメソッドを使わずにputsしています。

cellotak committed:
ls.rbのfile_detailとして拾ってくる情報を追加

cellotak committed:
ls.rbのoutputメソッドを-lオプションに対応させる

cellotak committed:
ls.rbで、出力時にprint文の羅列をせず変数にまとめていれるように変更

cellotak commented on 04.ls/ls.rb:
```diff
@@ -11,39 +11,70 @@ def parse_options(argv)
 end
 
-def fetch_file_names(target_directory_path, options)
+def fetch_file_details(target_directory_path, options)
```
このメソッドにおいて、ファイル名とその詳細情報をどういう形にするかの方針で悩んでいます。

いくつかプランがありますが、現在はプランAで書いてあります。

## プランA
以前はファイル名のみを配列の要素として持ったfile_namesを返すメソッドとしていたが、ファイル名やそのほか-lオプションで表示したい詳細情報をすべて一つのハッシュにまとめ(file_detial)、そのハッシュを配列の要素としてもつfile_detailsを返す。
```ruby
#構造の例
file_details = [ {name: "hoge" , mode: xxxxxx, :size xxxx, ... } ,  {name: "fuga" , mode: xxxxxx, :size xxxx, ... }, ... ]
````

### メリット
- ファイルに関する情報をすべてfile_detailsに一元化できる

### デメリット
- -l オプションがないときの挙動は変える必要がないが、その部分の実装も変えてしまうことになる。
- detailsという名前だが、-l オプションがないときは `:name`要素だけしか持たないので実情と合ってない感がする
    - これに関しては - l オプションなかろうが 詳細情報をfile_detailsに入れればいいのかもしれないが、使わないデータを変えさせるのは少し気持ちが悪い (この程度の内容なら無視してよさそうだが、仮にデータ的に重めの情報が含まれるシチュエーションだとよくなさそう)
- outputメソッドでの処理がわかりづらくなる
　以前はfile_names_tableだったものをfile_details_tableに変更しているが、file_details_tableという名前が混乱を招きそう。そして結局 `:name`しか出力しないので、ちぐはぐな印象
　
## プランB
file_namesはそのまま利用し、file_namesの各要素(つまりfile_name)をキーとしたハッシュを新たに作り、その値として詳細情報をハッシュにしたものを持たせる(つまりハッシュが入れ子になる)。

```ruby
#構造の例
file_names = ["hoge", "fuga", ...]
file_details = { "hoge" :  {mode: xxxxxx, :size xxxx, ...} ,  "fuga" : {mode: xxxxxx, :size xxxx ...}, ... }
````
出力するときはfile_namesに対して繰り返し処理をして、そこで出てきたfile_nameをキーとしてfile_detailsからハッシュを呼び出す、というイメージ
```ruby
 row_count.times do |row_index|
      col_count.times do |col_index|
        target_file_name = file_details_table[col_index][row_index]
        print file_details[target_file_name][:mode] 
        #以下省略
      end
      print "\n"
```

## メリット
- -l オプション無し時のoutputが以前と同様、file_namesをそのまま使えるので個人的には分かりやすく感じる
- file_namesがファイル名、file_detailsがファイル名以外の詳細情報をもっているので名前とデータの中身の実情が合っているように感じる

## デメリット
- 同じファイルに対して配列とハッシュが別々に定義されているのは気持ち悪いし、構造的にバグを生みやすそう。

## プランC
オブジェクト指向的に書く
オブジェクト指向的に書いた経験がないのでメリデメが曖昧。

### メリット
- 出力したい情報を持つインスタンスを持つようにクラスを作ればなんかスッキリかけそうなイメージ 。(どんな設計になるかあまり想像がついてはいない)

### デメリット
- コードが根本から変わってしまって大変そう
- オブジェクト指向がそこまで身についてないのでハマる可能性が大
- 裏事情ですが、実はしばらく先で ls コマンドをオブジェクト指向化するプラクティスがある([ls コマンドを作るオブジェクト指向版 \| FBC](https://bootcamp.fjord.jp/practices/213))ので、今回やってしまうとオブジェクト指向使わずに書いたときにどうなるかを感じる機会が失われそう


## プランD
Dataクラスを使う (森さんの提案の一つ)
Dataクラスを把握していないので、メリデメ考えられていません。

　
　
　
　

ebihara-tomonari replied:
ちょっと質問の意図とズレてしまうかもしれないのですが、
そもそも `fetch_file_names` を `fetch_file_details` に置き換えようとしている理由と、
`file_details` を作ろうとしている理由が分かりませんでした

一つの考え方としてなんですが、今回 `-l` オプションを付与されなかった時の振る舞いは全く変える必要ないはずなので、 `-l` オプションが渡されなかった時の既存実装の成果物をできれば弄りたくないです
根本的に作り変えるほどバグやデグレの元になりがちで、レビューや動作確認負担も増えるので、根本的な方針として今の方針で作業進めるのは避けたいかな、と思いました
そのため `fetch_file_names` メソッドの定義はそのまま残してて良い気がしますし、
`output` メソッドの方は確かに分岐必要になっちゃうんですが、これまでの定義の方を `output_default_formatt` メソッドみたいな命名に変えて差分を減らしたいです

```ruby
def output(file_names, options)
  return if file_names.empty?

  if options[:l]
    output_long_listing_format(file_names)
  else
    output_default_format(file_names)
  end
end

def output_default_format(file_names)
  row_count = ((file_names.size - 1) / MAX_COL_COUNT) + 1
  col_count = [file_names.size, MAX_COL_COUNT].min

  # 以下略。なるべく既存実装のまま
end

def output_long_listing_format(file_names)
  # 今回必要になった -l オプション付与時の出力定義だけ新規作成する
end
```

次に `file_details` の使い道とか意義なんですが、 `file_names` があれば各ファイルに対応する `File::Stat` を導き出せるはずなので、都度 `File::Stat` インスタンスを生成すれば良いかなとは思うんですよね
少なくとも今のコードみたいに、生成時に
```ruby
  if options[:l]
    file_details.each do |file_detail|
      stat = File.stat("#{target_directory_path}/#{file_detail[:name]}")
      file_detail[:mode] = stat.mode # 要変換
      file_detail[:hardlink_count] = stat.nlink
      file_detail[:owner_user] = stat.uid # 要変換
      file_detail[:owner_group] = stat.gid # 要変換
      file_detail[:size] = stat.size
      file_detail[:ctime] = stat.ctime
    end
  end
```
して、呼び出し時に
```ruby
  if options[:l]
    file_details.each do |file_detail|
      detail_str = 
        "#{file_detail[:mode]} "\
        "#{file_detail[:hardlink_count]} "\
        "#{file_detail[:owner_user]} "\
        "#{file_detail[:owner_group]} "\
        "#{file_detail[:size]} "\
        "#{file_detail[:ctime]} "\
        "#{file_detail[:name]}"
      puts detail_str
    end
```
と書いてあるのは、ほぼ同じことがコード内に二度登場しているような気がして読み辛さを生んでいる気がします
であれば、最初から
```ruby
file_names.each do |file_name|
  stat = File.stat("#{target_directory_path}/#{file_name}")
  detail_str = 
    "#{stat.mode} "\
    "#{stat.nlink} "\
    "#{stat.uid} "\
    "#{stat.gid} "\
    "#{stat.size} "\
    "#{stat.ctime} "\
    "#{file_name}"
  puts detail_str
end
```
的に書いてあった方が良いと思います

cellotak replied:
ちょっとまだその背景部分をまとめられていなかったのでわかりづらかったかと思いますが、返信する形で説明させてもらいます。

- ### `fetch_file_names` を `fetch_file_details` に置き換えようとしている理由

元々、`fetch_file_names` はoutputに渡すデータを用意するもの(この場合だとfile_names)、`output`は受け取ったデータを形だけ整えて出力するもの、という役割をそれぞれ持たせるつもりで書いていました。

-l オプションがついて表示する情報が多くなった時でも「`fetch_file_names` はoutputに渡すデータを用意するもの」、「`output`は受け取ったデータを形だけ整えて出力するもの」という部分は変えたくなかったので、受け取るデータの設計を変える必要があると考えました。


そこで`file_details`というデータの形をつくり、`File::stat`から情報を取ってきて`file_details`に落とし込むように`fetch_file_names`を置き換えていきました。

最終的に、扱っているものが`file_names`ではなく`file_details`になったのでメソッドの名前を`fetch_file_details`にしたという経緯です。
(うまく伝わってるでしょうか、、？)

元々`file_names`だった内容を`file_details`が内包しているように変えたのは、同じファイルに対して別々の配列が出来上がっているのがよくないと考えたから、というのが一番大きいです。

- ### `file_details` を作ろうとしている理由
上記部分と内容若干被りますが、もう少し書きます。
これも背景の説明がないのでわかりづらかったかと思いますが、現状`output`でputsしている値は仮のもので、`mode`とか`owner_user`とか`owner_group`はFile::Statからとってきたものから変換する必要があります。その変換部分をoutput内ではやらせたくないので、`file_details` を作ってそこによしなに変換したしたデータを入れていきたいという考えです。


- ### -lオプション無しの時の既存実装について
> 今回 -l オプションを付与されなかった時の振る舞いは全く変える必要ないはずなので、 -l オプションが渡されなかった時の既存実装の成果物をできれば弄りたくないです

デメリットとしてあげるのを忘れていましたが、これはかなり気がかりな点でした。file_detailsは作るにせよ、file_namesとは独立させた方がよさそうですね。ただその場合、やはり同じファイルに対して別の配列ができあがるのが気になりはします。

ebihara-tomonari replied:
> 元々、fetch_file_names はoutputに渡すデータを用意するもの(この場合だとfile_names)、outputは受け取ったデータを形だけ整えて出力するもの、という役割をそれぞれ持たせるつもりで書いていました。
> -l オプションがついて表示する情報が多くなった時でも「fetch_file_names はoutputに渡すデータを用意するもの」、「outputは受け取ったデータを形だけ整えて出力するもの」という部分は変えたくなかったので、受け取るデータの設計を変える必要があると考えました。

なるほど～
僕は output に渡す前処理で取り扱いたいファイル対象や順序を整理できれば良いんだろうな、くらいの感覚でしたね

> 次に file_details の使い道とか意義なんですが、 file_names があれば各ファイルに対応する File::Stat を導き出せるはずなので、都度 File::Stat インスタンスを生成すれば良いかなとは思うんですよね

の通りなんですが、
こと出力時の観点において `file_names` の順に処理させれば良いって点が変わらないのであれば、 `file_names` とは別に `file_details` の集合を作る必要性を依然持ってなくて、
`File.stat` から得られた値をなんやかんやする部分をロジック分離をさせたいなら、それは `file_details` を作るって形ではなく、
例えば一ファイルに対してその `-l` オプション形式の情報出すために `File.stat` を利用するって形にすると
```ruby
def output_long_listing_format(file_names)
  file_names.each do |file_name|
    puts long_listing_file_detail(file_name)
  end
end

def long_listing_file_detail(file_name))
  stat = File.stat("#{target_directory_path}/#{file_name}") # 簡略化コードなので target_directory_path は引数で渡すか、グローバルに参照できる変数になってるのか、上手いことやってください
  # なんか前処理を色々やる
  permissions = ......

  # 最後に出力する
  result = permissions + hardlink_count + ... + file_name
end
```
こんな感じにまとめられる方が良いかな、と思いました

とりあえず変更差分の影響度合いの多さや、今のぱっと見の印象のように冗長な処理っぽい流れのせいで行数が多くなりすぎると読むの辛いかな、と思うのでそこはなんとかしたいですね～

cellotak replied:
なるほど！確かによく考えたら`output`に`File::Stat`から得た値を変換する処理をさせたくないからといって、`file_details`という箱ものをわざわざ用意する必要はないですね、、、
値を変換するメソッドを`output`の中で1ファイルごとにを呼びだす、という発想は自分の中になかったので気づけませんでした。

「出力する」という処理を実装しようとするとき、つい箱を用意してそこに内容全部揃えてしまう癖がありますね。。(そうするべき時もあるかもしれませんが)

cellotak committed:
Revert "ls.rbを既存の実装を活かす方針に変更したので、ls3提出時まで戻す"

This reverts commit 296e0a47c3e4ac1631d967bb1ac74d813b0460c8.
file_namesをfile_detailsに置き換える方針だったが、file_namesを残す方針に変更した。

cellotak committed:
Revert "ls.rbのoutputメソッドを-lオプションに対応させる"

This reverts commit 7c6a1c5b9b116ce8633a5f0378fa4a2737e304d9.

cellotak committed:
Revert "ls.rbのfile_detailとして拾ってくる情報を追加"

This reverts commit c9ad23fce6837cde2117c39b9428ddc707fcc312.

cellotak committed:
Revert "ls.rbに追加する-lオプションのため、fetch_file_namesメソッドを変更"

This reverts commit 92cac5fd1120e4697c58db89c8405a04dc1b1c86.

cellotak committed:
ls.rbについて、outputメソッドを変更

-lオプションの有無によってoutputメソッドの中でさらに呼び出すメソッドを切り替えるように変更。-lオプション無しの場合に呼び出されるメソッドは変更前のoutputメソッドの実装とほぼ同じ

cellotak committed:
ls.rbについて、uidやgitを名前に変換するように変更

cellotak committed:
ls.rbにファイルタイプを文字列で表示できる機能を追加

cellotak committed:
ls.rbのパーミッションを表示できる機能を追加

cellotak committed:
ls.rbの日付の表示をOS標準コマンドに合わせた

cellotak committed:
ls.rbのrubocopによる修正とリファクタリング

cellotak committed:
ls.rbのconvert_permission_num_to_strをリファクタリング

cellotak commented on 04.ls/ls.rb:
```diff
@@ -2,28 +2,41 @@
+def output(file_names, target_directory_path, options)
   return if file_names.empty?
 
+  if options[:l]
```
オプションによる条件分岐が1つメソッドにまとまられていないのが気になりますが、現状これが素直な条件分岐だと感じます。

JunichiIto replied:
ここは特に違和感を感じなかったですー。

cellotak commented on 04.ls/ls.rb:
```diff
@@ -42,8 +55,48 @@ def output(file_names)
+end
+
+def convert_stat_mode_to_str(stat_mode)
+  file_type_num = format('%06o', stat_mode).slice(0..1)
+  file_type_str = FILE_TYPE_LIST[file_type_num]
+
+  permission_num = format('%06o', stat_mode).slice(3..5)
+  permission_str = convert_permission_num_to_str(permission_num)
+
+  file_type_str + permission_str
```
`num`や`str`という一般的な略称をつかっていますが、このあとの`convert_permission_num_to_str`メソッド内では複数形を使いたいがために`strings`のように略さず命名しています。
ここもそれに合わせて略さずに命名すべきか、もしくは複数形の方を`strs`にするか、どちらが適切でしょうか。

JunichiIto replied:
numだと整数値かと思いましたが、ここは文字列（数値ではなく数字）ですね。
numはcode、strはchar（1文字）みたいな名前にすると収まりがいいかなーと思いました。

cellotak replied:
codeやchatという変数名の案を使用させて頂きました。

cellotak commented on 04.ls/ls.rb:
```diff
@@ -42,8 +55,48 @@ def output(file_names)
+  file_type_str + permission_str
+end
+
+def convert_permission_num_to_str(permission_num)
+  permission_num_digits = permission_num.chars.map(&:to_i)
+  partial_permission_strings = permission_num_digits.map do |digit|
+    partial_permission_string = '-' * 3
+
+    partial_permission_string[0] = 'r' if digit / 4 == 1
+    partial_permission_string[1] = 'w' if (digit / 2).odd?
+    partial_permission_string[2] = 'x' if digit.odd?
+
+    partial_permission_string
+  end
+  partial_permission_strings.join
+end
```
このメソッド内の変数名も全体的に悩ましいです。
`ditit`という変数名はは3桁の数字を各桁に分解したときの単体の数字、という意味で使っていますが適切でしょうか。
パーミッションの記号表記の9文字をpermission_stringと定義していて、3文字ずつの塊を`partial_permission_string`, その塊を3つ集めた配列を`partial_permission_strings`と命名しましたが、これも自分では意味がわかるのですが、他人から見て理解できるのかが判断つきません。

JunichiIto replied:
こういうケースは自力で命名を頑張るより。英語版のmanや公式ドキュメントなどを読んでみて、「公式がそれを何と呼んでいるのか」を確認するのがオススメですー。

cellotak replied:
https://linuxconfig.org/understanding-of-ls-command-with-a-long-listing-format-output-with-permission-bits
↑こちらに`octets` という名前で説明されていたので、その名前を参照しました。

cellotak commented on 04.ls/ls.rb:
```diff
@@ -42,8 +55,48 @@ def output(file_names)
+  file_names.each do |file_name|
+    stat = File.stat("#{target_directory_path}/#{file_name}")
+
+    detail_str =
+      "#{convert_stat_mode_to_str(stat.mode)} "\
+      "#{stat.nlink} "\
+      "#{Etc.getpwuid(stat.uid).name} "\
+      "#{Etc.getgrgid(stat.gid).name} "\
+      "#{stat.size} "\
+      "#{stat.ctime.strftime('%b %-d %H:%M')} "\
+      "#{file_name}"
+    puts detail_str
```
この出力部分、いくつかの項目はメソッドチェインしてものを直接出力していますが、中身を一回変数に置き換えるべきか判断迷っています。

JunichiIto replied:
メソッドチェインっていうのは`Etc.getpwuid(stat.uid).name`や`stat.ctime.strftime('%b %-d %H:%M')}`のことでしょうか？
これぐらいなら許容範囲かなーと思います。

cellotak commented on 04.ls/ls.rb:
```diff
@@ -42,8 +55,48 @@ def output(file_names)
   end
 end
 
+def output_long_listing_format(file_names, target_directory_path)
+  file_names.each do |file_name|
+    stat = File.stat("#{target_directory_path}/#{file_name}")
+
+    detail_str =
+      "#{convert_stat_mode_to_str(stat.mode)} "\
+      "#{stat.nlink} "\
+      "#{Etc.getpwuid(stat.uid).name} "\
+      "#{Etc.getgrgid(stat.gid).name} "\
+      "#{stat.size} "\
+      "#{stat.ctime.strftime('%b %-d %H:%M')} "\
+      "#{file_name}"
+    puts detail_str
+  end
+end
```
現状のロジックですと1ファイルごとに情報を呼び出して出力するので、項目ごとに幅を揃えて出力することができません。(下記画像参照)
![image](https://github.com/cellotak/ruby-practices/assets/65953037/5056f93e-c4e6-45ca-b881-69877f3947b3)


項目ごとに幅揃えをする機能は必須でしょうか。

幅を揃えるとした場合、各項目の情報をすべて出してから、各項目の文字数の最大値を取得し、出力の時にその最大値を元に幅揃えして出力、というかなりややこしい実装になるかと思われます。

課題としてそこまで求められていなかった場合に、そこにあまり時間をかけすぎたくなかったので、幅揃えは未実装です。

JunichiIto replied:
動画内でもコメントしましたが、一度チャレンジしてみてほしいです！

cellotak replied:
幅揃え実装してみました。

cellotak reviewed (commented):
悩んでいるポイントや自分なりに判断して決めた部分をコメントで記載させて頂きました。
ご意見いただければ幸いです。

JunichiIto reviewed (commented):
動画でレビューしたので別途リンクを送ります〜。
ただ、質問は動画内で回答するのを忘れたのでコメントを返信しておきました。

cellotak committed:
トップレベルの処理をmainメソッドにまとめて先頭に移動

cellotak committed:
ls.rbのメソッドの記述箇所を移動

cellotak committed:
ls.rbでtotalを末尾に表示できる機能を追加

cellotak committed:
ls.rbの変数名やメソッド名を変更

cellotak committed:
ls.rbの文字列の破壊的変更があった部分を修正

cellotak committed:
ls.rbに桁揃えの機能追加

cellotak committed:
幅揃えの機能に関してリファクタリング

cellotak commented on 04.ls/ls.rb:
```diff
@@ -42,8 +111,28 @@ def output(file_names)
+  permission_octets = permission_code_nums.map do |num|
+    permission_octet_chars = []
+
+    permission_octet_chars << (num / 4 == 1 ? 'r' : '-')
+    permission_octet_chars << ((num / 2).odd? ? 'w' : '-')
+    permission_octet_chars << (num.odd? ? 'x' : '-')
```
@JunichiIto 
この部分ついて動画内で「自分で考えたんでしょうか？」コメントされていましたが、自分で考えたオリジナルのロジックになります。
一度2進数に変換してから文字列に変換するのが2度手間に感じたので、直接変換できる方法はないか考えた結果、このような条件で分岐できることがわかりました。

こういったロジックは可読性的に問題あるでしょうか？

JunichiIto replied:
なるほど、独自のロジックなんですね。あまり見たことがないロジックだったので、出典がちょっと気になってました😅

> こういったロジックは可読性的に問題あるでしょうか？

いえ、このあたりの処理はどうしてもややこしくなりがちなので、うまくメソッド分割できていれば問題ありません〜。

ebihara-tomonari replied:
【MAY】
個人的には二進数で構成されているのが分かっているのであれば、二進数っぽい処理しちゃう方がどう言う処理しているか分かりやすいかな、とは思いました
が、確かにRubyで二進数ならではっぽい処理を良い感じに書こうとするのどうするのが良いんでしょうね 🤔 

例えば
```ruby
PERMISSONS = %w(r w x).freeze

permission_to_binary = num.to_s(2).ljust(3, "0")
permission_to_binary.chars.map.with_index do |bit, i|
  bit.to_i.zero? ? "-" : PERMISSONS[i]
end.join
```
とか

あと、この辺は一桁ずつにどの項目に対してのパーミッションかどうかの意味があるので、その辺も可視化されてても嬉しいかなとは個人的に思いました

```ruby
def convert_permission_code_to_str(permission_code)
  owner, group, other = permission_code.chars.map(&:to_i)

  convert_permission_to_s(owner) +
    convert_permission_to_s(group) +
    convert_permission_to_s(other)
end

def convert_permission_to_s(permission_code_num)
  # ここは一桁の数字を 'rw-' みたいな文字列に変換する処理
  # 上述の二進数処理に該当するはず
end
```

って感じで
一見、元々のコードよりも出力の仕方が周りくどくなったかもですが、
出力時の各桁数をフラットに処理するのではなく、各桁の意味を抑える形のロジックの書かれ方になるので、よりビジネスロジックらしさが出ると言うか `-l` オプションで出力されている内容が分かっていればいるほど、対応しているコード部分が分かりやすい可読性が出て来る気はします

cellotak commented on 04.ls/ls.rb:
```diff
@@ -2,28 +2,97 @@
+def output_long_listing_format(file_names, target_directory_path)
+  block_count_total = 0
+
+  details_by_file_name = {}
```
これは `file_name`をキー、`detail`を配列でまとめた`details`をvalueとしてもつハッシュです。
`file_names`をeachで回して`file_name`に対応した`details`を呼び出せるようにこのような構造にしています。

cellotak committed:
幅揃えの機能に関して再度リファクタリング

cellotak commented on 04.ls/ls.rb:
```diff
@@ -2,28 +2,78 @@
+  end
+end
+
+def output_long_listing_format(file_names, target_directory_path)
```
rubocopでこのメソッドを20行以内にするように指摘されましたが、これ以上行数を削るには、さらに処理をメソッドに分割する以外の方法がわかりません。
これ以上メソッドへ分割して入れ子が深くなってもわかりづらくなるような気がするのですが、どうするべきでしょうか。

JunichiIto replied:
> これ以上行数を削るには、さらに処理をメソッドに分割する以外の方法がわかりません。
これ以上メソッドへ分割して入れ子が深くなってもわかりづらくなるような気がするのですが、どうするべきでしょうか。

なるほど〜。
ではいったんcellotakさんなりのメソッド分割をやってみてもらえますか？
その結果を見て、わかりづらい場合は改善案をコメントさせてもらいます。

rubocopの指摘はパスさせることが必須条件なので、がんばってパスさせてみてください。
（もちろん、むりやり感を出さずにスマートに実装できればベストです）

cellotak replied:
メソッド分割してみました。
意外とわかりやすくなったかもしれません。
ただ、入れ子がより深くなっていたり、メソッドの引数をそのままさらに別のメソッドに渡したりしているので、引数に対して変更があった場合に影響範囲が大きく、保守性が悪いのではないかと懸念しています。

cellotak reviewed (commented):
指摘して頂いた内容の修正と、幅揃えの機能を追加しました。
意図がわかりづらそうな部分や、疑問点をコメントしています。
ご確認お願い致します。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -9,6 +9,14 @@
+# directory_pathsには複数のpathを指定することは許容しているが、現時点でファイル名を表示するのは1番目に指定したディレクトリのみにしている。
+def main
+  options, directory_paths = parse_options(ARGV)
+  target_directory_path = directory_paths[0] || './'
```
```suggestion
  # directory_pathsには複数のpathを指定することは許容しているが、現時点でファイル名を表示するのは1番目に指定したディレクトリのみにしている。
  directory_path = directory_paths[0] || './'
```

コメントを書くなら対象のコードに一番近いところがいいと思います。

あと、「ターゲットじゃないdirectory_path」があるわけではないので、`target_`は削って良さそうです

cellotak replied:
コメントの位置を変更しました。

確かにおっしゃる通り、ターゲットじゃないものはないのでかえってややこしい表現ですね。
削りました！

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -42,8 +92,50 @@ def output(file_names)
+  details << stat.size.to_s
+  details << stat.ctime.strftime('%b').to_s
+  details << stat.ctime.strftime('%-d').to_s
+  details << stat.ctime.strftime('%H:%M').to_s
```
102-104はまとめてフォーマットした方が効率がいいと思います。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -42,8 +92,50 @@ def output(file_names)
-output(file_names)
+def convert_stat_to_details(stat)
+  details = []
+  details << convert_stat_mode_to_str(stat.mode).to_s
```
文字列に対して`to_s`を呼ぶのは無駄にメソッド呼び出しを増やしてしまうだけなのでなくしましょう。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -42,8 +92,50 @@ def output(file_names)
+  details << stat.ctime.strftime('%H:%M').to_s
+end
+
+def fetch_max_widths_by_detail(details_by_file_name)
```
fetchはネットワーク上のサーバーやデータベースなど、プログラム本体の外にあるリソースを取得するときに使うことが多いです。
というわけで別の名前を検討してみてください〜。

cellotak replied:
なるほど、そういったニュアンスがあるのですね。
`get`, `build`, `create`、`extract` などの候補を考えてみましたが、元々データとしてあるものを取得する動きがメインになるのかなと考え、`get`に変更してみました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,28 +2,78 @@
+    details = details_by_file_name[file_name]
+    details.each_with_index do |detail, index|
+      if /^\d+$/.match?(detail)
+        print detail.rjust(max_widths_by_detail[index])
```
配列の二人三脚をやってるので、もっと堅牢なロジックにしたいですね。
[【プログラミング初心者向け】配列の二人三脚を避けよう \- Qiita](https://qiita.com/jnchito/items/19c4b2c1c6690ebe8f25)

cellotak replied:
二人三脚するのではなく、それぞれの繰り返しの中で幅揃えするためのwidthを求めるようにしました。

JunichiIto replied:
`details` と `max_widths_by_detail` は必ず同じ要素数かつ同じ並びでなければならない、という意味で、二人三脚は脱してないですね。
配列からハッシュへの変更をお勧めします〜。

cellotak committed:
mainメソッドを修正

cellotak committed:
mainメソッドを修正

cellotak committed:
convert_stat_to_detailsメソッドをリファクタリング

cellotak committed:
メソッド名を変更

cellotak committed:
配列の二人三脚にならないように変更

cellotak committed:
output_long_listing_formatの動作をさらにメソッドに分割

cellotak committed:
メソッドの記述順を変更

cellotak reviewed (commented):
再度修正してみました。
レビューお願い致します

JunichiIto reviewed (commented):
コメントしました！

> ただ、入れ子がより深くなっていたり、メソッドの引数をそのままさらに別のメソッドに渡したりしているので、引数に対して変更があった場合に影響範囲が大きく、保守性が悪いのではないかと懸念しています。

プログラミングの一般論ですが、データを作成処理とデータを画面に出力する処理でさらにメソッドを分けてあげると保守性が良くなるかもしれません。

cellotak committed:
details_by_file_nameの中身を配列ではなくハッシュに変更

cellotak commented on 04.ls/ls.rb:
```diff
@@ -2,28 +2,129 @@
+
+  file_names.each do |file_name|
+    details = details_by_file_name[file_name]
+    details_output_order = %i[stat_mode nlink username groupname size ctime]
+
+    details_output_order.each do |key|
+      width = get_max_width_by_detail(details_by_file_name, key)
+      if /^\d+$/.match?(details[key])
+        print details[key].rjust(width)
+      else
+        print details[key].ljust(width)
+      end
+      print ' '
+    end
+    puts file_name
+  end
+end
+
```
`details_by_file_name`のvalueを配列からハッシュに変更しました。
意図した順序で出力するために、出力したい順序にkeyを並べた配列を作りました。

cellotak reviewed (commented):
前回レビューの際、ローカルの変更がGithubに反映されていなかったかもしれません。
申し訳ありません。

file_changedを見ながら返信しているつもりだったのですが、ローカルのコードを見ながら返信を書いてしまっていたようで、未反映なことに気づいていませんでした。

いずれにせよ、「`details`という配列が同じ要素数かつ同じ並びでなければならない、という意味で、二人三脚は脱してない」状態だったので、ハッシュを使ってみて修正してみました。

レビューお願い致します。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,28 +2,129 @@
+
+def calc_block_count_total(file_names, target_directory_path)
+  block_count_total = 0
+  file_names.each do |file_name|
```
sumメソッドを使うとシンプルに書けます

cellotak replied:
sumメソッドを使うように変更しました。
一行で書くこともできましたが、わかりづらそうだったので、mapメソッドで配列にして、endの後にsumメソッドを付ける形にしました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,28 +2,129 @@
+  file_names.each do |file_name|
+    stat = File.stat("#{target_directory_path}/#{file_name}")
+    details = convert_stat_to_details(stat)
+    details_by_file_name[file_name] = details
```
eachをto_hに変えて、

```suggestion
    [file_name, details]
```

とするとよりシンプルに書けます

cellotak replied:
to_hの使い方、あまりわかっていませんでしたが、こういったところで使えるのですね。
to_hに変更しました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,28 +2,129 @@
+
+def convert_stat_to_details(stat)
+  details = {}
+  details[:stat_mode] = convert_stat_mode_to_str(stat.mode)
```
```suggestion
  {
    stat_mode: convert_stat_mode_to_str(stat.mode),
```

好みの問題ですが、ハッシュリテラル一発で書いてしまう手もあります。

cellotak replied:
その書き方が頭から抜けていました。
`details[:key]`と何回も書くのを避けたいので、提案どおり修正させて頂きました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,28 +2,129 @@
+end
+
+def get_max_width_by_detail(details_by_file_name, key)
+  detail_widths_by_file_name = details_by_file_name.map { |_file_name, details| details[key].length }
```
```suggestion
def get_max_width_by_detail(details_by_file_name, key)
  sleep 0.1
  detail_widths_by_file_name = details_by_file_name.map { |_file_name, details| details[key].length }
```

こんな感じでsleepを入れたりするとめちゃくちゃ遅くなりませんか？
無駄なメソッドを呼び出すを減らすように検討してみてください〜。

cellotak replied:
何度も呼び出す処理をやめて、最大幅のデータもハッシュ化することでkeyで呼べるようにしました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,28 +2,129 @@
+  block_count_total
+end
+
+def get_details_by_file_name(file_names, target_directory_path)
```
get_ って戻り値を返すメソッド全般に使えてしまうのと、Rubyがあまり get_ や set_ という動詞を使わない、という点から、get_ 以外の名前（より具体的な動詞）を与えた方がベターです。

cellotak replied:
`build`に変更してみましたが、いかがでしょうか？

別の箇所(最大幅を算出するメソッド)でも`get`を使っていましたが、そちらは`calc`に変更しました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,28 +2,129 @@
+
+    details_output_order.each do |key|
+      width = get_max_width_by_detail(details_by_file_name, key)
+      if /^\d+$/.match?(details[key])
```
各項目にkeyが付いたので、正規表現よりもkey名で条件分岐する方がロジックとして堅牢になります。
（正規表現だと、意図しない入力値が入ってきたときに誤作動するリスクがついて回る）

cellotak replied:
なるほど、安易に正規表現は使わないようにします。
今回は右寄せする要素の key のリストを定数化して、それに当てはまるかどうかで条件分岐するように変更しました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,28 +2,129 @@
+
+  file_names.each do |file_name|
+    details = details_by_file_name[file_name]
+    details_output_order = %i[stat_mode nlink username groupname size ctime]
```
静的な配列なので定数にするのも一手です。

cellotak replied:
定数化しました。

JunichiIto reviewed (commented):
コメントしました〜。

cellotak committed:
メソッド名を修正

cellotak committed:
繰り返しで足し算していた処理をsumメソッドに変更

cellotak committed:
引数の名前を変更(target_directoryのtargetを削除)

cellotak committed:
ハッシュへの値代入の記述方法を変更

cellotak committed:
-lオプションの詳細情報出力順を指定する配列を定数化

cellotak committed:
右寄せか左寄せの条件分岐に正規表現を使わないようにし、用意されたkeyのリストに当てはまるかどうかで判別するように変更

cellotak committed:
ハッシュの生成をeachメソッドからto_hメソッドに変更

cellotak committed:
-lオプションで表示する詳細の文字列の最大値を求めるメソッドが繰り返し呼び出されていた問題を解消

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,28 +2,130 @@
+end
+
+def calc_block_count_total(file_names, directory_path)
+  file_names.map do |file_name|
```
map + sumではなく、sumだけで同じ結果を得ることができます。

cellotak replied:
`Enumerable#sum`の方だったのですね！
つい`Array#sum`の方を参照していました 😅

JunichiIto reviewed (approved):
確認しました！OKです〜🙆‍♂️

ebihara-tomonari commented on 04.ls/ls.rb:
```diff
@@ -2,28 +2,130 @@
 
+FILE_TYPE_LIST = { '01' => 'p', '02' => 'c', '04' => 'd', '06' => 'b', '10' => '-', '12' => 'l', '14' => 's' }.freeze
+DETAILS_OUTPUT_ORDER = %i[stat_mode nlink username groupname size ctime].freeze
+RJUST_LIST = %i[nlink size].freeze
```
【MAY】
前回も似たような話出た気がするけど最終的にRJUSTさせる、って末端の具体的処理が命名の基準の癖になるとあまり良くない気はしてて、何故RJUSTさせることになる要素なのかって方にフォーカスしてくれた方が対象物の本質に近いので、
`NUMERIC_LOMG_LISTING_ITEMS` みたいな命名でも良いかもですね

cellotak replied:
命名の際に具体的な処理ではなく、対象物の本質的なところを基準にする、というのを気を付けます。

ebihara-tomonari commented on 04.ls/ls.rb:
```diff
@@ -2,28 +2,130 @@
+  max_width_by_detail = calc_max_width_by_detail(details_by_file_name)
+
+  file_names.each do |file_name|
+    details = details_by_file_name[file_name]
```
各ファイル名に対するパーミッションとかの詳細データのことを `detail` 
その集まりのことを `details` と呼んでいる気がするので、ここではあるファイルに対する詳細データ一つだけっぽいので `detail` が適切かな？

それとも `max_width_by_detail` の命名的には `stat_mode` , `nlink` みたいなものを `detail` と呼んでいて、それらの集まりであるハッシュだからそれは `details` として取り扱ってます？
なんかこの辺の用語の使い方が怪しいかも？

cellotak replied:
`stat_mode` , `nlink` みたいなものを `detail` と呼んでいます。なので1つのファイルに対して`details`が紐づいていて、それらをさらに集めたハッシュが `details_by_file_name`としています。
でも実際説明書いてみると、detailsとdetails_by_file_nameの関係性がわからないですね。

`stat_mode` , `nlink` みたいなものを`detail_element`などにしてみようかと思いました。

ebihara-tomonari commented on 04.ls/ls.rb:
```diff
@@ -2,28 +2,130 @@
+      else
+        print details[key].ljust(max_width_by_detail[key])
+      end
+      print ' '
```
ここ最後の項目に対してもスペース出力して問題ないのかって思ったけど、最後にファイル名を出力しつつ改行から問題ないのか

【MAY】
個人的には、順々に print , puts を呼び出したりするよりは、一行分の出力をちゃんと作って間と間をつなげる半角スペースは `join` とかで用意する処理に仕上げる方が洗練されている感が出そうな気はしました

```ruby
line_output =
  DETAILS_OUTPUT_ORDER.map do |key|
    if RJUST_LIST.include?(key)
      details[key].rjust(max_width_by_detail[key])
    else
      details[key].ljust(max_width_by_detail[key])
    end
  end.push(file_name).join(' ')

puts line_output
```

この方が print を複数回読ぶ必要はなく、一行に付き一回で出力処理が完結しつつ、半角スペースの付与をループ内でやらずにちゃんと一行分の出力内容を作るって過程で処理できるので、
少し綺麗な処理の仕方をしている感が出る気がします

cellotak replied:
なるほど、まとめてしまえばいいんですね！

ebihara-tomonari commented on 04.ls/ls.rb:
```diff
@@ -2,28 +2,130 @@
+
+    permission_octet_chars.join
+  end
+  permission_octets.join
```
【MAY】
レビュー過程でいくつか既に削られているとは思いますが、
そう言えば、中間処理の成果物の変数が多いせいで、頑張ってそれぞれに変数名付けないといけなくなってるのも名付け辛さの要員かもですね
プラス処理中の意味合いが大きくなるほど普通に意味が分かり易そうな概念の命名を付けづらくなってるかも？

例えば、この辺は処理の方向性は変えなくても、
```ruby
def convert_permission_code_to_str(permission_code)
  permission_code.chars.map do |num_s|
    num = num_s.to_i
    read_permisson = num / 4 == 1 ? 'r' : '-'
    write_permisson = (num / 2).odd? ? 'w' : '-'
    execute_permisson = num.odd? ? 'x' : '-'

    read_permisson + write_permisson + execute_permisson
  end.join
end
```
とかにしちゃっても良いかも？

cellotak replied:
なるほどー、そうやって書くとかなり読みやすいですね
メソッドをつなげすぎてわかりづらくなることを恐れて1個ずつ中間生成物的なものに名前をつけていましたが、それはそれで下手にやると命名が微妙になってわかりづらくなるんですね。

ebihara-tomonari reviewed (commented):
お疲れさまでした！
修正不要ですが、せっかくなので一応気になった点など軽くフィードバックしておきます～
